test/validation/api/ipsec/ipsec_inline_in.c

```diff
@@ -10,6 +10,10 @@ static int ipsec_sync_init(odp_instance_t *inst)
 {
 	int rc;
 
+	suite_context.inbound_op_mode = ODP_IPSEC_OP_MODE_INLINE;
+	suite_context.outbound_op_mode = ODP_IPSEC_OP_MODE_ASYNC;
+	suite_context.reass_ipv4 = true;
```

Comment:
Maybe the name of this function (ipsec_sync_init) could be fixed (ipsec_init?) now that you are modifying it otherwise too.

Reply:
Another issue is that this adds reassembly testing only for the inline case. Maybe the testing should be enabled for the async case too, which would require some changes in packet handling in the async inbound processing code.

Reply:
> Maybe the name of this function (ipsec_sync_init) could be fixed (ipsec_init?) now that you are modifying it otherwise too.

Well, ipsec_init() doesn't convey meaning correctly either, right?. I see the same issue is there with all other test init functions. I would like to not mix that with current set of changes. I can make such a change (if you can suggest a name that is less ambiguous) but I would like to keep that change not squashed to reassembly UT changes.

>  Another issue is that this adds reassembly testing only for the inline case. Maybe the testing should be enabled for the async case too, which would require some changes in packet handling in the async inbound processing code.

Wouldn't it be better to enable it when we have some implementation which can actually go through those code paths? I would expect even async code can leverage the current proposed routines, but some breaking down of existing functions would be required. But isn't it better attempted then?

Reply:
> Well, ipsec_init() doesn't convey meaning correctly either, right?

It performs initialization and the ipsec seems to be a prefix generally used in this code. But actually the function cannot be renamed ipsec_init() as such function already exists and is called by this very function.

My point was that ipsec_sync_init name is wrong because this does not have anything to do with synchronous mode and it would be ok to fix the naming now that your are touching this code anyway. But if you do not want to do the rename now, then it is ok to leave it as it is as the naming issue is not otherwise really related to this PR.

> Wouldn't it be better to enable it when we have some implementation which can actually go through those code paths?

Maybe so, if the test implementations significantly differ. But I do not think there should be any difference since the reassembly test case logic should not care or need to know whether async or inline operations are used.

Reply:
@JannePeltonen @psavol 

> But if you do not want to do the rename now, then it is ok to leave it as it is as the naming issue is not otherwise really related to this PR.

Agreed. I see that ipsec_async.c has ipsec_async_init() & ipsec_sync.c has ipsec_sync_init(). Making the one in inline to ipsec_init() would leave the code even less uniform. I believe this we can take up in a separate PR.

> Maybe so, if the test implementations significantly differ. But I do not think there should be any difference since the reassembly test case logic should not care or need to know whether async or inline operations are used.

It matters. The datapath has fundamentally different pieces for ASYNC & INLINE. Probably can you check the patch which adds reassembly of IPv4. The functions being called there describes the operations.

1. Take individual fragments
2. Perform IPsec outbound processing
3. Verify all such IPsec processed fragments. Note that we have N IPsec processed packets now.
4. Send this to pktio interface
5. Reassembly happens during inline Rx path
6. Verify the received packet for further reassembly handling. --> Only one odp_queue_dequeue() applicable for this.

With proposed changes, 1-4 constitutes ipsec_out_verify_send() and 5-6 constitutes ipsec_in_recv_verify(). With ASYNC, steps 4 & 5 need to be skipped. How you do that is something I chose not to answer now. Basically there are 2 options,

1. Add a check for reassembly.inline_mode in the above functions() --> Current approach in the overall UT. I'm not a fan of this approach. I had also proposed something similar in a previous version which you had opposed. 
2. Break the existing functions and call the required ones from a newly introduced function. --> Scalable option in IMO.

Given the lack of implementations supporting this, the code added cannot be tested. So I would like to leave these ideas open for overall community and implementations that would support these features.

For now, I would like to focus on inline reassembly paths. Extending it to support ASYNC can be taken up as a separate PR, right? I intend to move the reassembly tests into a separate suite and then add it only in ipsec_inline_in to make this more obvious.

Reply:
> The datapath has fundamentally different pieces for ASYNC & INLINE

What I have been trying to preserve is the separation of the IPsec processing mechanism (sync, async,inline) and the test case logic. Previously the processing mechanism was hidden inside ipsec_check_out() and ipsec_check_in_one() and the test cases did not need to know what operation mode was configured.

Now the reassembly tests introduce ipsec_out_verify_send() and ipsec_in_recv_verify() that work only in inline mode and restrict the scope of the validation tests, which is not ideal.

I think the needed adjustment for reassembly in the event reception could be added in a way that they apply for both inline and async cases. The only real difference between async and inline event reception is the handling of non-matching packets, i.e. the pktio queue dequeueing part. If the common part, i.e. the normal event reception, of those two were first combined to a shared function, then the reassembly related changes would automatically apply to the async case too.

Another change could be to have test_part::num_pkt really mean the expected number of output packets as I think it originally did (it was even called out_pkt) and fix the sending & reception code that currently does not seem to be able to decide what num_pkts is for. Once this is done, the reassembly test code could just use ipsec_check_out (or ipsec_process_out) and ipsec_check_in_one just like everybody else (which zero expected number of output packets until the reassembled packets is expected).

Reply:
@JannePeltonen If I wasn't clear earlier, the scheme you are suggesting would mean, 

1. test_part::pkt_in will be an array of pointers and we will need to introduce another field to denote the number of packets.
2. There will be suite_context.reass checks in the **"existing"** functions

If you agree to both the above (second one especially), then I'll proceed in that direction. Hope you realize that we are making a full u-turn from what was confirmed on the cleanup PR. I'll drop all the cleanup patches in this PR as these don't add any value if test-cases aren't making use of them.

> Now the reassembly tests introduce ipsec_out_verify_send() and ipsec_in_recv_verify() that work only in inline mode and restrict the scope of the validation tests, which is not ideal.

Only other option is to add a check for suite_context.reass in the existing function(something similar was part of the earlier version, which you had opposed). Please do confirm if this is fine.

Reply:
> test_part::pkt_in will be an array of pointers and we will need to introduce another field to denote the number of packets.

Not necessarily. That is needed only if you want to be able to call ipsec_check_in_one/ipsec_process_in_one only once, giving it a set of input packets and an expected output packet. But that does not enough in case of incomplete reassembly when you get multiple output packets.

Another way is to still have only one input packet in the operations but N output packets instead of always one. In reassembly tests you would first call ipsec_process_in_one() with a number of ESP packets without expecting any output packet. And then you would call ipsec_process_in_one() with the final packet and expecting the reassembled result packet. So in the first calls num_pkt would be 0 and in the last call it would be 1.

> There will be suite_context.reass checks in the "existing" functions

I am not sure I can follow. Basically you already added such check in an existing function by adding it in the inline event reception even though that was factored out to its own function.

I do not quite see why there even needs to be any suite_context.reass checks. I guess the event processing code could simply check if it received non-IPsec subtype packet and then check reassembly status etc. Or there could be suitable flag in test part to better tell what kind of output packet is expected.

Reply:
@JannePeltonen 

May be let's talk about the test case. This talking from top level is not heading anywhere.

```
static void test_in_ipv4_esp_reass_success_two_frags(odp_ipsec_sa_t out_sa,
	
						     odp_ipsec_sa_t in_sa)
{
	ipsec_test_part test_out[MAX_FRAGS], test_in[MAX_FRAGS];
	part_prep_esp(test_in, 2, false);
	
	test_in[0].pkt_in = &pkt_ipv4_udp_p4_f1;
	test_in[1].pkt_in = &pkt_ipv4_udp_p4_f2;
	ipsec_out_verify_send(test_in, out_sa, 2);
		
	part_prep_plain(test_out, 1, 1, false, true);
	test_out[0].out[0].pkt_res = &pkt_ipv4_udp_p4;

	ipsec_in_recv_verify(test_out, in_sa, 1);
	
}
```
Do you have any concerns on following line?
`
part_prep_esp(test_in, 2, false);
`

Next line,
`ipsec_out_verify_send(test_in, out_sa, 2);`

The existing function which does similar is
`ipsec_check_out()`

Difference is `ipsec_check_out()` doesn't send out packets. It does outbound processing and verifies the packet. We can loop in N times to process N frags, but we will need some mechanism in the caller to convert the processed odp_packet_t into ipsec_test_packet.

Hope you don't have any concerns on the following line
`part_prep_plain(test_out, 1, 1, false, true);`

Next line,
`ipsec_in_recv_verify(test_out, in_sa, 1);`

Can be replaced with,
`ipsec_check_in_one()` if we can give N test_part to this, with first N-1 would be send for just sending out packets, and last packet will have one packet to send and one packet to receive.

Basically an edit like,
```
-               CU_ASSERT_EQUAL(1, recv_pkts_inline(part, pkto));
+               if (part->num_pkt)
+                       CU_ASSERT_EQUAL(part->num_pkt, recv_pkts_inline(part, pkto));
        }
```
in ipsec_process_in_one()

If the above is fine, I think I can add the below code. This would meet your requirements in using existing code.

```
static void test_in_ipv4_esp_reass_success_two_frags(odp_ipsec_sa_t out_sa,
						     odp_ipsec_sa_t in_sa)
{
	ipsec_test_part test_out[MAX_FRAGS], test_in[MAX_FRAGS];
	ipsec_test_packet test_pkt[MAX_FRAGS];
	odp_packet_t pkt[MAX_FRAGS];
	int i, j;

	memset(test_in, 0, sizeof(ipsec_test_part) * MAX_FRAGS);

	part_prep_esp(test_out, 2, false);
	test_out[0].pkt_in = &pkt_ipv4_udp_p4_f1;
	test_out[1].pkt_in = &pkt_ipv4_udp_p4_f2;

	for (i = 0; i < 2; i++) {
		CU_ASSERT_EQUAL(ipsec_check_out(&test_out[i], out_sa, &pkt[i]), 1);
		ipsec_prep_part_from_pkt(&test_in[i], &pkt[i], &test_pkt[i]);
	}

	part_prep_plain(&test_in[i - 1], 1, 1, false, true);
	test_in[i-1].out[0].pkt_res = &pkt_ipv4_udp_p4;

	for (j = 0; j < i; j++)
		ipsec_check_in_one(&test_in[j], in_sa);
}
```

IMO, the code above looks very cumbersome to figure out what is being attempted and how is it getting tested. But then, it is your call. Please confirm on the above. If the above sequence can work, I'll make changes accordingly and submit.

Reply:
> part_prep_esp(test_in, 2, false);

This is ok (except for what I commented separately about making sure that test_in array is always large enough).

> ipsec_out_verify_send(test_in, out_sa, 2);

One issue I have with this is that it works only with inline inbound processing mode and even implies that in the name. Another issue is that the function does two somewhat separate things (converting plain text packets to ESP packets through outbound IPsec processing and then feeding the ESP packets as input to the actual inbound processing that is under testing).

> ipsec_check_in_one() if we can give N test_part to this, with first N-1 would be send for just sending out packets, and last packet will have one packet to send and one packet to receive.

I am thinking more in the lines of having ipsec_check_in_one() take one test part like it does now but make it work also when the number of expected output packets is zero.

The fragment reassembly test would then have these steps:
1) Take the plain text fragment packets and do outbound processing for them so that we get ESP packets for the actual resssembly tests.
2) call ipsec_check_in_one() with the first fragment and expect no packet out
3) call ipsec_check_in_one() with the second packet and expect the reassembled packet out.

I think things are clearer when step is separate from the later steps. The first step is not even needed in potential other test cases that have the ESP packets defined in test vectors.

The separation helps also if one wants to add e.g. timing related testing (checking that the implementation really waits for subsequent fragments and does not give up too ealy) or mangle the packets a bit before inbound processing.

> IMO, the code above looks very cumbersome to figure out what is being attempted and how is it getting tested.

That may be partly because of how it is written and partly because it does not use any helper functions like the other version (ipsec_out_verify_send). I am not sure the other version is truly clearer since you have to know what exactly what ipsec_out_verify_send() is doing and how it interacts with recv_pkts_inline().

Reply:
So your suggestion is to split the below such that first we will have a loop sending packets (and not expecting anything back) and then final ipsec_check_in_one() call with a packet to send and receive.

```
	for (j = 0; j < i; j++)
		ipsec_check_in_one(&test_in[j], in_sa);
```

Like ,

```
	for (j = 0; j < i-1; j++)
		ipsec_check_in_one(&test_in[j], in_sa);

        /* Update test_in[i-1] */

        ipsec_check_in_one(&test_in[i-1], in_sa);
```

Also, can you confirm if the change in ipsec_check_in_one() path is acceptable? Also, please do suggest on the code snippet above. If that would work for you, I'll rework rest of the cases as well.

Reply:
> So your suggestion is to split the below such that first we will have a loop sending packets (and not expecting anything back) and then final ipsec_check_in_one() call with a packet to send and receive.

Yes, something like that. It does not necessarily have to be two loops like that, the code could also interleave the outbound and inbound processing. And in case of just two fragments, having for loops may just make the code less readable. I would just like to build on top of the existing processing functions and not force coupling between outbound and inbound processing.

> Also, can you confirm if the change in ipsec_check_in_one() path is acceptable

I think the patch is for ipsec_process_in_one() as it is in the PR. The change looks ok for the inline path. Corresponding fix for the async path would look something like this:

```
-               pkt = ipsec_packet(part->pkt_in);
-               num_out = odp_ipsec_in_enq(&pkt, 1, &param);
-               CU_ASSERT_EQUAL(1, num_out);
+               int consumed;

-               num_out = (num_out == 1) ?  1 : 0;
+               pkt = ipsec_packet(part->pkt_in);
+               consumed = odp_ipsec_in_enq(&pkt, 1, &param);
+               CU_ASSERT_EQUAL(1, consumed);
+               if (consumed <= 0)
+                       num_out = 0;
```

Reply:
@JannePeltonen Okay. I tried redoing the code a bit. Please check if the below code snippet would work?

```
static void test_in_ipv4_esp_reass_success_four_frags(odp_ipsec_sa_t out_sa,
						      odp_ipsec_sa_t in_sa)
{
	ipsec_test_part test_out[MAX_FRAGS], test_in[MAX_FRAGS];
	ipsec_test_packet test_pkt[MAX_FRAGS];
	odp_packet_t pkt[MAX_FRAGS];
	int i;

	memset(test_in, 0, sizeof(ipsec_test_part) * MAX_FRAGS);

	CU_ASSERT(MAX_FRAGS >= 4);

	part_prep_esp(test_out, 4, false);

	test_out[0].pkt_in = &pkt_ipv4_udp_p5_f1;
	test_out[1].pkt_in = &pkt_ipv4_udp_p5_f2;
	test_out[2].pkt_in = &pkt_ipv4_udp_p5_f3;
	test_out[3].pkt_in = &pkt_ipv4_udp_p5_f4;

	part_prep_plain(&test_in[3], 1, 1, false, true);

	for (i = 0; i < 4; i++) {
		CU_ASSERT_EQUAL(ipsec_check_out(&test_out[i], out_sa, &pkt[i]), 1);

		ipsec_prep_part_from_pkt(&test_pkt[i], &pkt[i]);
		test_in[i].pkt_in = &test_pkt[i];

		ipsec_check_in_one(&test_in[i], in_sa);
	}
}
```

Reply:
Yes, it looks like it would work, but I would still propose a couple of minor changes:

- Change the last argument of memset to sizeof(test_in)
- Move test_pkt and pkt inside the loop, where they can be normal variables instead of arrays.
- Rename ipsec_prep_part_from_pkt to ipsec_test_packet_from_pkt or similar now that you changed it to not operate on test parts anymore.

Reply:
So can you confirm if I can proceed with the following template?

```
static void test_in_ipv4_esp_reass_success_four_frags(odp_ipsec_sa_t out_sa,
						      odp_ipsec_sa_t in_sa)
{
	ipsec_test_part test_out[MAX_FRAGS], test_in[MAX_FRAGS];
	int i;

	memset(test_in, 0, sizeof(test_in[0]) * MAX_FRAGS);

	CU_ASSERT(MAX_FRAGS >= 4);

	part_prep_esp(test_out, 4, false);

	test_out[0].pkt_in = &pkt_ipv4_udp_p5_f1;
	test_out[1].pkt_in = &pkt_ipv4_udp_p5_f2;
	test_out[2].pkt_in = &pkt_ipv4_udp_p5_f3;
	test_out[3].pkt_in = &pkt_ipv4_udp_p5_f4;

	part_prep_plain(&test_in[3], 1, 1, false, true);

	for (i = 0; i < 4; i++) {
		ipsec_test_packet test_pkt;
		odp_packet_t pkt;

		CU_ASSERT_EQUAL(ipsec_check_out(&test_out[i], out_sa, &pkt), 1);

		ipsec_test_packet_from_pkt(&test_pkt, &pkt);
		test_in[i].pkt_in = &test_pkt;

		ipsec_check_in_one(&test_in[i], in_sa);
	}
}
```

Reply:
Yes, but I would still change the last argument of memset to sizeof(test_in).

Reply:
Agreed. Also, I missed setting pkt_res in the earlier sequence. Updating both,

```
static void test_in_ipv4_esp_reass_success_four_frags(odp_ipsec_sa_t out_sa,
						      odp_ipsec_sa_t in_sa)
{
	ipsec_test_part test_out[MAX_FRAGS], test_in[MAX_FRAGS];
	int i;

	memset(test_in, 0, sizeof(test_in));

	CU_ASSERT(MAX_FRAGS >= 4);

	part_prep_esp(test_out, 4, false);

	test_out[0].pkt_in = &pkt_ipv4_udp_p5_f1;
	test_out[1].pkt_in = &pkt_ipv4_udp_p5_f2;
	test_out[2].pkt_in = &pkt_ipv4_udp_p5_f3;
	test_out[3].pkt_in = &pkt_ipv4_udp_p5_f4;

	part_prep_plain(&test_in[3], 1, 1, false, true);
	test_in[3].out[0].pkt_res = &pkt_ipv4_udp_p5;

	for (i = 0; i < 4; i++) {
		ipsec_test_packet test_pkt;
		odp_packet_t pkt;

		CU_ASSERT_EQUAL(ipsec_check_out(&test_out[i], out_sa, &pkt), 1);
		ipsec_test_packet_from_pkt(&test_pkt, &pkt);
		test_in[i].pkt_in = &test_pkt;

		ipsec_check_in_one(&test_in[i], in_sa);
	}
}
```

Please confirm if this looks good.

Reply:
Looks good to me.

Reply:
> Does the above assume odp_ipsec_in_enq() would not include packets for reassembly? If yes, the above is not correct, right?

I did not understand this comment and I do not see what the problem would be. The problem to be fixed in the existing code is that it confuses the expected number of output packets (num_out) and the number of consumed input packets. From now on the number of input packets and thus the number of packets consumed by odp_ipsec_in_enq() is still one, but the number of output packets to be dequeued may be one or zero.

Reply:
If you don't see any problem, I have no concerns.